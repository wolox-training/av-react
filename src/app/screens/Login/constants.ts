export const LOGIN_FIELDS = {
  email: 'email',
  password: 'password'
};
