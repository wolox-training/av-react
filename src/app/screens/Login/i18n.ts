import i18next from 'i18next';

i18next.addResources('es', 'Login', {
  logoAlt: 'WoloxLogo',
  email: 'Email',
  password: 'Password',
  signUp: 'Sign up',
  login: 'Login',
  required: 'Hay campos incompletos',
  invalidEmail: 'El formato de mail no es correcto'
});
