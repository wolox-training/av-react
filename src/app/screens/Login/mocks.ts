export const OK_LOGIN_API_RESPONSE = {
  data: {
    'access-token': 'test',
    data: {
      email: 'test@test.com',
      firstName: 'test',
      id: 1,
      lastName: 'test',
      locale: 'es'
    }
  }
};
