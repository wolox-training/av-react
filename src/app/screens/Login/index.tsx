import React from 'react';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';
import { Link, useHistory } from 'react-router-dom';

import { User, UserRequestSuccess } from '~utils/types';
import { login } from '~services/UserService';
// eslint-disable-next-line import/namespace
import { useLazyRequest } from '~app/hooks/useRequest';
import Loading from '~components/Spinner/components/loading';
import CustomErrorDisplayer from '~components/CustomErrorDisplayer';
import { PATHS } from '~constants/paths';
import { CLIENT_KEY, TOKEN_KEY, UID_KEY } from '~utils/constants';
import LocalStorageService from '~services/LocalStorageService';
import CustomInput from '~components/CustomInput';

import WoloxImg from '../Assets/wolox-logo.png';

import { LOGIN_FIELDS } from './constants';
import styles from './styles.module.scss';

export default function Login() {
  const history = useHistory();

  const loginSuccess = (data?: UserRequestSuccess) => {
    if (data) {
      LocalStorageService.setValue(TOKEN_KEY, data.accessToken);
      LocalStorageService.setValue(UID_KEY, data.uid);
      LocalStorageService.setValue(CLIENT_KEY, data.client);
      history.push(PATHS.home);
    }
  };

  const { register, handleSubmit, errors } = useForm<User>({ mode: 'all' });
  const [, loading, error, sendRequest] = useLazyRequest({
    request: login,
    withPostSuccess: loginSuccess
  });
  const onSubmit = handleSubmit(data => sendRequest(data));
  return (
    <div className="column center">
      <form className={`column ${styles.loginForm}`} onSubmit={onSubmit}>
        <img src={WoloxImg} alt={i18next.t('Login:logoAlt') as string} className={styles.logo} />
        <CustomInput
          name={LOGIN_FIELDS.email}
          inputRef={register({
            required: { value: true, message: i18next.t('Login:required') },
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: i18next.t('Login:invalidEmail')
            }
          })}
          error={errors.email?.message}
          label={i18next.t('Login:email') as string}
          labelClassName={styles.customLabel}
          type="email"
        />
        <CustomInput
          name={LOGIN_FIELDS.password}
          inputRef={register({
            required: { value: true, message: i18next.t('Login:required') }
          })}
          error={errors.password?.message}
          label={i18next.t('Login:password') as string}
          labelClassName={styles.customLabel}
          type="password"
        />
        <div className={`column ${styles.loginButtonContainer}`}>
          <button className={`${styles.customButton} ${styles.loginButton}`} disabled={loading} type="submit">
            {i18next.t('Login:login')}
          </button>
        </div>
        <Link to={PATHS.signup} className="row">
          <button
            className={`${styles.customButton} ${styles.signupButton} full-width`}
            type="button"
            disabled={loading}
          >
            {i18next.t('Login:signUp')}
          </button>
        </Link>
      </form>
      {loading && <Loading className={styles.loading} />}
      {error?.errorData?.errors.length && <CustomErrorDisplayer errors={error.errorData?.errors} />}
    </div>
  );
}
