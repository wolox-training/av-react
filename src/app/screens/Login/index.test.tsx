import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';

import Login from './';

const MAX_BUTTONS_LENGTH = 2;
const MAX_ALERTS_LENGTH = 2;

describe('Login test', () => {
  let container: Element | null = null;

  beforeEach(() => {
    const { container: renderContainer } = render(
      <Router>
        <Login />
      </Router>
    );
    container = renderContainer;
  });

  test('Should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  test('Should render email and password inputs', () => {
    expect(screen.getByLabelText('Login:email')).toBeInTheDocument();
    expect(screen.getByLabelText('Login:password')).toBeInTheDocument();
  });

  test('Should render login and signup buttons', () => {
    expect(screen.getAllByRole('button').length).toBe(MAX_BUTTONS_LENGTH);
  });

  test('Should render required message after submitting form with empty fields', async () => {
    userEvent.click(screen.getByText('Login:login'));
    await waitFor(() => expect(screen.getAllByRole('alert').length).toBe(MAX_ALERTS_LENGTH));
  });
});
