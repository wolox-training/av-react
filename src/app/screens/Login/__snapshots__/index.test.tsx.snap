// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login test Should match snapshot 1`] = `
<div>
  <div
    class="column center"
  >
    <form
      class="column loginForm"
    >
      <img
        alt="Login:logoAlt"
        class="logo"
        src="wolox-logo.png"
      />
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="email"
        >
          Login:email
        </label>
        <input
          class="input full-width"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="password"
        >
          Login:password
        </label>
        <input
          class="input full-width"
          id="password"
          name="password"
          type="password"
        />
      </div>
      <div
        class="column loginButtonContainer"
      >
        <button
          class="customButton loginButton"
          type="submit"
        >
          Login:login
        </button>
      </div>
      <a
        class="row"
        href="/sign_up"
      >
        <button
          class="customButton signupButton full-width"
          type="button"
        >
          Login:signUp
        </button>
      </a>
    </form>
  </div>
</div>
`;
